### PR TITLE
Addes typed dependencies to the lib package

### PR DESCRIPTION
**Pull request by ajsantander**
Solves #478 

This PR introduces typed dependencies for the lib package, which can be divided into 3 groups:
1) Dependencies like axios or bignumber.js, that already included `.d.ts` files, for which nothing needed to be done. Note that Web3 v1.x is typed in this way, but Web3 v0.x (the one we're using) is not. 
2) Dependencies like ethereumjs-abi, for which custom `.d.ts` files were written in src/@types.
3) Dependencies like Web3 v0.x which don't have official declaration files in the DefinitelyTyped repository, but unofficial declaration files were found.

In short, the most interesting part of this PR is that we are using typings for Web3 v0.x from `web3-typescript-typings`. The typings were used only in packages/lib/src/artifacts/ZWeb3.ts but could ripple up to a lot of code above, such as Transactions.ts and much, much more. This would be very interesting, since it would define a large part of the interfaces we use, that are currently being used as `any`.

### CODE REVIEW

File: packages/lib/src/@types/ethereumjs-abi.d.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+declare module 'ethereumjs-abi' {
```
[the remark this answers is not in the document]
Comment (ajsantander) [reply packages/lib/src/@types/ethereumjs-abi.d.ts]: These aren't generated files, I created them manually. I believe that it's common for Typescript project to include a folder like this one for stuff that is not typed in dependencies.

File: packages/lib/src/artifacts/BuildArtifacts.ts [hunk 2]
```diff
@@ -47,11 +47,7 @@ export class BuildArtifacts {
   }
 
   public registerArtifactForSourcePath(sourcePath: string, artifact: Artifact): void {
-
-    if (!this.sourcesToArtifacts[sourcePath]) {
-      this.sourcesToArtifacts[sourcePath] = [];
-    }
-
+    if (!this.sourcesToArtifacts[sourcePath]) this.sourcesToArtifacts[sourcePath] = [];
```
[the remark this answers is not in the document]
Comment (ajsantander) [reply packages/lib/src/artifacts/BuildArtifacts.ts]: Atm you can do this:

```
if (foo.aLargeFunctionNameThatCanAnnoyMeInASingleLine(3, 'foo', 42) 
  bar = foo.anotherLargeFunctionNameThatCanAnnoyMeInASingleLine()
```

TSLint, by itself is not smart enough to require curlies for single argument conditionals that are a bit long. However, I think that there is an ESLint plugin for TSLint that could handle this.

File: packages/lib/src/artifacts/ContractFactory.ts [hunk 3]
```diff
@@ -1,11 +1,12 @@
 import BN from 'bignumber.js';
 import ZWeb3 from './ZWeb3';
+import Web3 from 'web3';
 import decodeLogs from '../helpers/decodeLogs';
 import { getSolidityLibNames, hasUnlinkedVariables } from '../utils/Bytecode';
 
 interface ContractSchema {
   contractName: string;
-  abi: any;
+  abi: Web3.AbiDefinition[];
```
[the remark this answers is not in the document]
Comment (ajsantander) [reply packages/lib/src/artifacts/ContractFactory.ts]: If we update to Web3 1.x, quite a lot of things would change I think. Web3's interface would change, including how it uses promises instead of callbacks. I think that in this case, having some sort of interface would help when that moment comes, because we would be looking for similar structures.

File: packages/lib/src/project/AppProject.ts [hunk 4]
```diff
@@ -39,13 +39,9 @@ export default class AppProject extends BasePackageProject {
...
         ? await App.fetch(appAddress, txParams)
         : await App.deploy(txParams);
-      if (packageAddress) {
-        thepackage = await Package.fetch(packageAddress, txParams);
-      } else if (await app.hasPackage(name, version)) {
-        thepackage = (await app.getPackage(name)).package;
-      } else {
-        thepackage = await Package.deploy(txParams);
-      }
+      if (packageAddress) thepackage = await Package.fetch(packageAddress, txParams);
```
[the remark this answers is not in the document]
Comment (ajsantander) [reply packages/lib/src/project/AppProject.ts]: Yes agreed. We could try the ESLint plugin I mentioned above.

File: packages/lib/src/artifacts/ContractFactory.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jcarpanelli) [reply packages/lib/src/artifacts/ContractFactory.ts]: I'm concerned about having to modify these custom types because of external decisions (i.e., external lib updates). I prefer leaving all these variables as `any`, but I don't have a strong reason to support my thoughts.

**Comment by ajsantander** [comment]
Putting this PR on hold until we have a better understanding of wether we want to type depenencies like web3 manually.
